# (연습) 함수

**참고**

먼저 [함수](https://codingalzi.github.io/code-workout-python/functions/) 내용을 공부한 다음에 연습문제를 풀어볼 것을 권장한다.

## 함수 호출과 표현식

### 연습문제

**문제 1**

영어 알파벳에서 소문자와 대문자의 유니코드 코드 포인트(정수값) 차이는 항상 일정하게 유지된다.
이를 이용하여, 변수 `char_upper`에 할당된 대문자 `'A'`의 코드 포인트를 구한 후에
특정 숫자를 더해 소문자 `'a'` 로 변환시키는 코드를 작성하라. 

답:

먼저 두 알파벳의 유니코드 코드 포인트의 차이를 계산한다.
소문자 알파벳의 유니코드 코드 포인트가 더 큼에 주의한다.

In [ ]:
char_upper = 'A'
diff = ord('a') - ord('A')  # 소문자와 대문자의 코드값 차이 (32)
print('대문자 A와 소문자 a의 코드값 차이:', diff)

대문자 A와 소문자 a의 코드값 차이: 32


대문자 유니코드 코드 포인트에 32를 더하면 소문자 유니코드 코드포인트가 계산된다.
이값에 `chr()` 함수를 적용하면 소문자로 변환된 알파벳이 구해진다.

In [ ]:
char_lower_code = ord(char_upper) + diff
char_lower = chr(char_lower_code)

print("대문자 A를 소문자로 변환:", char_lower)

대문자 A를 소문자로 변환: a


**문제 2**

영어 알파벳 `'a'`와 `'z'`의 유니코드 코드 포인트(정수값)의 평균을 구하면, 
알파벳 순서상 한가운데 위치한 문자가 무엇인지 수학적으로 알아낼 수 있다.
`ord()` 함수를 이용해 `'a'`와 `'z'`의 코드 포인트를 더한 뒤 2로 나눈 몫(정수 나눗셈 `//`)을 구하고, 
그 결과값을 `chr()` 함수에 전달하여 정중앙에 있는 문자를 확인하는 코드를 작성하라.

답:

두 유니코드 코드 포인트의 평균값을 두 정수의 합을 2로 나눈 몫으로 계산하고,
계산된 정수에 해당하는 알파벳을 확인하면 된다.

In [ ]:
code_a = ord('a')
code_z = ord('z')

# 두 코드 포인트의 더한 후 2로 나눈 몫 (평균값 구하기)
mid_code = (code_a + code_z) // 2

# 계산된 코드 포인트의 문자 출력
print("a와 z의 한가운데 문자:", chr(mid_code))

a와 z의 한가운데 문자: m


**문제 3**

ROT13 암호화를 영어 대문자 알파벳에 대해 작동하도록 구현한 다음에 대문자 알파벳 `'C'`와 `'P'`가 암호화된 알파펫을 확인하라.

힌트: 대문자 알파벳 `'A'`와 `'Z'`의 유니코드 코드 포인트를 확인한다.

답:

먼저 대문자 알파벳 `'A'`와 `'Z'`의 유니코드 코드 포인트를 확인한다.

In [ ]:
ord('A')

65

In [ ]:
ord('Z')

90

영어 대문자 알파벳의 유니코드 범위는 65(`'A'`)부터 90(`'Z'`)이다.
원래 문자의 유니코드 코드 포인트에 13만큼 더했을 때 대문자의 마지막인 `'Z'`(90)를 넘어간다면, 
알파벳 전체 개수인 26을 빼주어 다시 맨 앞(`'A'`)으로 순환하도록 만들어 준다. 

- `ord(char) + 13 <= 90`인 경우: `ord(char)`에 13을 더해준 값에 해당하는 알파벳
- `ord(char) + 13 > 90`인 경우: `ord(char)`에 13을 더한 뒤 26을 뺀 (결과적으로 13을 빼준) 값에 해당하는 알파벳

아래 코드가 이 성질을 이용하여 대문자 알파벳 `'C'`를 암호화 시킨다.

In [ ]:
char = 'C'

if ord(char) + 13 <= 90:
    encrypted = chr(ord(char) + 13)
else:
    encrypted = chr(ord(char) - 13)

print("원래 문자:", char)
print("암호화된 문자:", encrypted)

원래 문자: C
암호화된 문자: P


## 사용자 정의 함수

### 연습문제

**문제 1**

정수 `num`을 인자로 해서 호출되었을 때, 1의 자릿수가 3의 배수이면 입력값의 세 배를, 
아니면 두 배를 반환하는 함수 `three_or_not()` 함수를 구현하라.
매개변수는 `num` 하나만 사용한다.

힌트: `if ... else ...` 조건문과 나머지 연산자 `%` 활용

**답**

함수를 정의하기 위해 먼저 주어진 조건으로부터 정의되어야 하는 함수의 시그니처와 반환값이 다음과 같음을 알 수 있다.

- 시그니처: `def three_or_not(num):`
- 반환값: `num * 3` 또는 `num * 2`

반환값은 인자로 주어진 정수의 1의 자릿수가 3의 배수인지 여부에 따라 달라진다.
따라서 정수의 1의 자릿수를 확인하는 방법을 확인한 다음에 반환값을 지정하도록 
함수의 본문을 작성한다.

- 정수의 1의 자릿수 확인: 10으로 나눴을 때의 나머지

In [1]:
123 % 10

3

In [2]:
27 % 10

7

- 3의 배수 여부: 3으로 나눈 나머지가 0인지 여부

In [3]:
3 % 3 == 0

True

In [4]:
7 % 3 == 0

False

종합하면, `num`을 10으로 나눴을 때의 나머지에 대해 3의 배수 여부를 확인하고
그에 따라 `num`의 3배수 또는 2배수를 반환값으로 지정하는
본문으로 함수를 구현하면 된다.

In [5]:
def three_or_not(num):
    unit = num % 10
    if unit % 3 == 0:
        return num * 3
    else:
        return num * 2

- 3은 3은 배수

In [6]:
three_or_not(13)

39

- 7은 3의 배수 아님

In [7]:
three_or_not(27)

54

**문제 2**

짝수, 홀수 여부를 판정하는 함수를 활용하는 문제다.

(1) 정수 인자와 함께 호출되어 인자가 짝수일 때는 1, 홀수일 때는 0을 반환하는 함수 `even_odd()`를 구현하라.
매개변수는 `num` 하나만 사용한다.
단, 함수 본문에서 `return 표현식` 명령문은 한 번만 사용되어야 한다.

힌트: `if ... else ...` 조건문과 나머지 연산자 `%` 활용

**답**

`num`의 인자로 지정되는 정수의 짝수, 홀수 여부는 2로 나눈 나머지가 0인지, 1인지에 의해 결정된다.
따라서 `num % 2 == 0` 논리식의 진릿값에 따른 `if` 조건문을 아래처럼 사용한다.

In [8]:
def even_odd(num):
    if num % 2 == 0:
        return 1
    else:
        return 0

In [9]:
even_odd(10)

1

In [10]:
even_odd(11)

0

위 함수는 하지만 `return 표현식` 명령문을 한 번만 사용해야 한다는 조건을 위배한다.
이유는 `if 조건문`을 사용하면서 각각의 경우에 대해 반환값을 지정하였기 때문이다.
따라서 조건 결과에 따라 매번 반환값을 바로 지정하는 대신
해당 경우에 반환될 값을 변수에 할당해서 저장해 둔다.
그런 다음 `if` 조건문이 완료된 후에 한 번만 해당 변수가 가리키는 값을
반환값으로 지정한다.
단, 각각의 경우에 정의되는 변수 이름이 동일해야 함에 주의한다.

In [11]:
def even_odd(num):
    if num % 2 == 0:
        answer = 1
    else:
        answer = 0  
      
    return answer

In [12]:
even_odd(10)

1

In [13]:
even_odd(11)

0

(2) `even_odd()` 함수를 이용하여 인자가 짝수면 2로 나눈 몫을,
홀수이면 3배 더하기 1을 한 값을 반환하는 `mini_collatz()` 함수를 정의하라.
참고로 `collatz`는 콜라츠라고 읽으며,
콜라츠는 [콜라츠 추측](https://ko.wikipedia.org/wiki/%EC%BD%9C%EB%9D%BC%EC%B8%A0_%EC%B6%94%EC%B8%A1)이라는 
아직 미해결로 남아 있는 문제를 1937년에 제시한 독일 수학자 이름이다.

답:

주어진 값 `num`이 짝수인지 여부는 `num % 2 == 0`으로 판단하여
`mini_collatz()` 함수를 다음과 같이 정의할 수 있다.

In [14]:
def mini_collatz(num):
    if num % 2 == 0:
        return num // 2
    else:
        return num * 3 + 1

위 정의는 하지만 `even_odd()` 함수를 사용하지 않기에 수정이 필요하다.
이미 정의된 함수를 이용하려면 해당 함수의 시그니처, 특히 인자에 대한 정보와
반환값의 정보를 알아야 한다.

`even_odd()` 함수는 인자로 주어진 정수의 짝수, 홀수 여부에 따라 1 또는 0을 반환한다.
이 두 성질을 이용하여 `mini_collatz()` 함수를 다음과 같이 정의할 수 있다.
특히, `num % 2 == 0`을 `even_odd(num) == 1`로 대체함에 주의한다.

In [15]:
def mini_collatz(num):
    if even_odd(num) == 1:
        return num // 2
    else:
        return num * 3 + 1

- 짝수는 2로 나누기

In [16]:
mini_collatz(10)

5

- 홀수는 3배 더하기 1

In [17]:
mini_collatz(7)

22

## `None` 반환값

### 연습문제

**문제 1**

아래 코드를 실행하면 한 줄의 문자열이 정상적으로 출력된 후 오류가 발생한다.
화면에 출력되는 문자열이 무엇인지 확인하고, 오류가 발생하는 이유를 설명하라.

In [ ]:
def double_number(n):
    result = n * 2
    print("두 배 계산 완료:", result)

value = double_number(5)
final_value = value + 10
print("최종 값:", final_value)

두 배 계산 완료: 10


TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

답:

코드를 실행하면 제일 먼저 아래 문자열이 정상적으로 출력된다.

    두 배 계산 완료: 10

하지만 그 직후 함수 본문에 반환값을 지정하지 않은 `double_number()` 함수가 호출되어
`final_value`에 사칙연산에 사용할 수 없는 `None`이 할당된다.
최종적으로 연산에 사용할 수 없는 자료형의 값을 사용해서 발생한 오류를 의미하는
`TypeError`가 발생하게 된다.

**문제 2**

아래 코드의 실행 결과를 설명하라.

In [ ]:
def funcA() :  
    print("A")

def funcB() :  
    print("B")
    funcA()   

funcB()

B
A


답:

코드는 윗줄의 명령문부터 차례대로 실행된다.
따라서 아래 순서대로 명령문이 실행된다.

- `funcA()`, `funcB()` 두 개의 함수 정의. 모두 인자 사용하지 않음.
- `funcB()` 호출: 
    - `print("B")` 실행. 따라서 `"B"` 출력 
    - `funcA()` 호출: 
        - `print("A")` 실행. 따라서 `"A"` 출력
            
따라서 아래 순서대로 화면에 출력된다.

    B, A

**문제 3**

아래 코드의 실행 결과를 설명하라.

In [ ]:
print("Hi")   

def funcA() :  
    print("A")

def funcB() :  
    print("B")
    funcA()   

def funcC() :
    print("C")
    funcA()
    funcB()

print("D")
funcC()

Hi
D
C
A
B
A


답:

코드는 윗줄의 명령문부터 차례대로 실행된다.
따라서 아래 순서대로 명령문이 실행된다.

- `print("Hi")` 실행: `'Hi'` 출력
- `funcA()`, `funcB()`, `funcC()` 세 개의 함수 정의. 모두 인자 사용하지 않음.
- `print("D")` 실행: `"D"` 출력
- `funcC()` 호출
    - `print("C")` 실행. 따라서 `"C"` 출력
    - `funcA()` 호출: 
        - `print("A")` 실행. 따라서 `"A"` 출력
    - `funcB()` 호출: 
        - `print("B")` 실행. 따라서 `"B"` 출력 
        - `funcA()` 호출: 
            - `print("A")` 실행. 따라서 `"A"` 출력
            
따라서 아래 순서대로 화면에 출력된다.

    Hi, D, C, A, B, A

## 위치 인자와 키워드 인자

### 연습문제

**문제 1**

문자열과 양의 정수를 인자로 받아, 문자열을 정수만큼 반복 출력하는 함수
`text_copy_2or3()`를 구현하려 한다.
단, 매개변수는 다음 두 개를 사용하며, 시그니처는 다음과 같아야 한다.

```python
def text_copy_2or3(text, count=1):
```

- `text`: 첫째 인자. 문자열을 인자로 사용.
- `count`: 둘째 인자. 양의 정수를 인자로 사용. 기본 키워드 인자로 1 지정.
    0이하의 정수이면 `text` 인자를 그대로 화면에 출력해야 함.

예를 들어 `text_copy_2or3("Hello ", 3)`를 실행하면 다음처럼 출력돼야 한다.

```
Hello Hello Hello 
```

반면에 `text_copy_2or3("Hello ")`를 실행하면 다음처럼 출력돼야 한다.

```
Hello 
```

힌트: `print()` 함수 활용

답:

`text`에 주어진 문자열을 `count` 정수만큼 반복하려면 문자열 복제 연산을 이용한다.

```python
text * cound
```

복제 연산 결과를 출력하는 `print()` 명령문을 작성하면 된다.
단, 반환값에 대한 요구사항은 없기에 `return 표현식` 명령문은 사용하지 않는다.
즉, 함수의 반환값이 인자와 무관하게 항상 `None`이다.

In [ ]:
def text_copy_2or3(text, count=1):
    print(text * count)

In [ ]:
text_copy_2or3("Hello ", 3)

Hello Hello Hello 


In [ ]:
text_copy_2or3("Hello ")

Hello 


그런데 문자열 이어붙이기 연산은 양의 정수에 대해서만 제대로 작동한다.
0 이하의 정수에 대해서는 연산 결과가 무조건 빈 문자열이기 때문이다.
아래 두 코드를 실행하면 아무 것도 출력되지 않지만 사람 눈에 보이지 않는 빈 문자열 `''`이 출력된다.

In [ ]:
text_copy_2or3("Hello ", 0)

In [ ]:
text_copy_2or3("Hello ", -2)

아래 함수는 `count`가 양의 정수가 아닌 경우를 별도로 다뤄 첫째 인자를 그대로 화면에 출력하도록 한다.

In [ ]:
def text_copy_2or3(text, count=1):
    if count <= 0:
        print(text)
    else:
        print(text * count)

양의 정수에 대해서는 이전과 동일하게 작동한다.

In [ ]:
text_copy_2or3("Hello ", 3)

Hello Hello Hello 


In [ ]:
text_copy_2or3("Hello ")

Hello 


반면에 1보다 작은 값에 대해서는 첫째 인자를 다시 화면에 출력한다.

In [ ]:
text_copy_2or3("Hello ", 0)

Hello 


In [ ]:
text_copy_2or3("Hello ", -2)

Hello 


**문제 2**

문자열 또는 리스트 포함된 항목을 인덱싱으로 확인할 수 있다.
예를 들어 `'python'` 문자열의 0번 인덱스의 문자는 `'p'`이고,
`[1, 3, 5, 7, 9]` 리스트의 -2번 인덱스의 항목은 `7`이다.

In [ ]:
language = 'python'
odd_upto_10 = [1, 3, 5, 7, 9]

print('python 문자열의 0번 인덱스 항목:', language[0])
print('홀수 리스트의 뒤에서 2번째 항목:', odd_upto_10[-2])

python 문자열의 0번 인덱스 항목: p
홀수 리스트의 뒤에서 2번째 항목: 7


문자열 또는 리스트의 길이가 n일 때 인덱스로 허용되는 정수의 범위는 아래와 같다.

- 양의 정수 인덱스: 0부터 (n-1) 까지의 양의 정수
- 음의 정수 인덱스: -1부터 -n 까지의 음의 정수

즉, 아래 범위에 포함된 정수들만 인덱스로 사용 가능하다.

```python
-n <= 정수인덱스 < n
```

언급된 범위를 범위를 벗어나는 인덱스를 사용하면 오류가 발생한다.

예를 들어, 아래 두 코드는 허용 범위를 벗어난 정수를 인덱스로 사용하면
오류가 발생함을 보여준다.

In [ ]:
language[6]

IndexError: string index out of range

In [ ]:
odd_upto_10[-8]

IndexError: list index out of range

문자열 또는 리스트의 인덱싱과 동일한 기능을 지원하지만 허용된 범위의 인덱스가 아닐 때
오류를 발생시키는 대신 지정된 값을 반환하는 
`getitem()` 함수를 정의하라.
단, 함수 시그니처는 다음과 같아야 한다.

```python
def getitem(str_or_list, idx, default=None):
```

- `str_or_list`: 문자열 또는 리스트를 받는 매개변수
- `idx`: 인덱스를 받는 매개변수
- `default=None`: `idx`가 허용된 범위를 벗어날 때 오류를 발생시키는 대신 사용되는 반환값 지정. 기본 키워드 인자는 `None`

답:

`getitem(str_or_list, idx, default)` 형식의 함수 호출이 발생했을 때 함수의 본문은 아래 내용을 실행해야 한다.

1. `str_or_list`의 길이 확인
2. `idx`가 허용 범위에 속하는지 여부 판단
3. 반환값 지정
   - 인덱스가 허용된 범위에 속하는 경우: `str_or_list[idx]`를 통해 해당 위치의 항목 반환
   - 인덱스가 허용된 범위를 벗어난 경우: `default` 값 반환

위 내용을 반영하여 `getitem()` 함수를 다음과 같이 정의한다.

In [ ]:
def getitem(str_or_list, idx, default=None):
    # 인덱스가 허용된 범위 안에 있는지 확인
    if -len(str_or_list) <= idx < len(str_or_list):
        return str_or_list[idx]
    else:
        return default

허용된 범위에 속하는 정수 인덱스는 정상적으로 작동한다.

In [ ]:
getitem(language, 0)

'p'

In [ ]:
getitem(odd_upto_10, -2)

7

허용된 범위를 벗어난 정수 인덱스에 대해서는 `default` 값이 반환된다.

- `default`의 기본 키워드 인자 사용하는 경우

In [ ]:
getitem(language, 6)

- `default`의 키워드 인자를 별도로 지정하는 경우

In [ ]:
getitem(odd_upto_10, -8, '인덱스 허용범위 초과')

'인덱스 허용범위 초과'

## 지역 변수와 전역 변수

### 연습문제

**문제 1**

아래 코드를 실행할 때 발생하는 오류를 설명하라.

In [ ]:
def square_return(number):
    square = number ** 2
    return square

number = 17
square17 = square_return(number)

print(number, "의 제곱을 계산한", square17, "과", square, "는 동일한 값을 가리킨다.")

NameError: name 'square' is not defined

답:

위 코드 마지막 줄의 `print()` 함수의 다섯번 째 인자로 사용된 `square` 변수가
정의되지 않았다는 의미의 `NameError` 오류가 발생했다.
이유는 `square` 변수는 `square_return(number)`이 호출되어 `square_return()` 함수가
인자 17과 함께 실행될 때에만 존재 의미가 있는데 이미 반환값이 정해지고 함수가 종료되었기 때문이다.

**문제 2**

다음 코드를 실행한 후 화면에 출력되는 `a`와 `b`의 값을 예측하고, 
그렇게 출력되는 이유를 매개변수와 지역 변수의 특징에 기반하여 설명하라.

In [ ]:
a = 10
b = 20

def swap(a, b):
    temp = a
    a = b
    b = temp
    return a, b

swap(30, 40)
print("a =", a)
print("b =", b)

a = 10
b = 20


답:

함수 `swap()` 정의에 사용된 매개변수 `a`와 `b`는 함수 내부에서만 존재하는 지역 변수다. 
함수를 호출할 때 전달된 30과 40은 지역 변수 `a`와 `b`의 값을 바꾸는 데 사용될 뿐, 
함수 바깥에서 정의된 전역 변수 `a = 10`과 `b = 20`에는 아무런 영향을 주지 않는다. 
따라서 마지막 `print()` 함수는 원래의 전역 변수 값을 그대로 출력한다.

**문제 3**

아래 `add_item()` 함수는 장바구니 역할을 하는 리스트 전역 변수 `cart`에 새로운 물건을 추가하려고 작성되었다. 
하지만 코드를 실행하면 오류가 발생한다.

In [ ]:
cart = ["사과", "우유"]

def add_item(item):
    cart = cart + [item]

add_item("파이썬 책")
print("현재 장바구니:", cart)

UnboundLocalError: cannot access local variable 'cart' where it is not associated with a value

발생하는 오류의 원인을 설명하고, 매개변수와 반환값을 활용하는 방식으로 코드를 정상 작동하도록 수정하라.
단, 새로운 함수의 시그니처는 다음과 같아야 한다.

```python
def add_item(current_cart, item):
```

답:

함수 내부에서 할당 연산자(`=`)를 사용해 `cart` 변수에 값을 저장하려고 시도하는 순간, 파이썬은 `cart`를 그 함수의 지역 변수로 취급한다.
그러나 할당을 위한 계산식 `cart + [item]`에서 지역 변수 `cart`의 값이 아직 지정되지 않았기 때문에,
해당 변수를 확인할 수 없다는 오류(`UnboundLocalError`)가 발생한다. 

전역 변수에 의존하기보다는, 기존 장바구니 상태와 새로운 항목을 인자로 받아 새로운 장바구니 리스트를 반환하는 형식으로 수정하는 것이 좋다.

In [ ]:
cart = ["사과", "우유"]

def add_item(current_cart, item):
    new_cart = current_cart + [item]
    return new_cart

cart = add_item(cart, "파이썬 책")
print("현재 장바구니:", cart)

현재 장바구니: ['사과', '우유', '파이썬 책']


## `input()` 함수

### 연습문제

**문제 1**

사용자로부터 입력받은 정수에 대한 3의 배수 여부를 판정하는 코드를 작성하라.
단, 3의 배수인 경우와 아닌 경우 각각 다르게 아래 형식으로 출력되도록 한다.

    24 은(는) 3의 배수다.
    
또는

    25 은(는) 3으로 나누면 나머지가 1 이다.

답:

3의 배수 여부는 3으로 나눈 나머지에 의존한다.
나머지가 0일 때만 3의 배수며, 나머지 연산자를 이용하여 확인한다.

아래 코드는 24를 입력할 때의 결과를 보여준다.

In [ ]:
user_input = int(input("정수 하나를 입력하세요: "))
print("입력한 값:", user_input)

if user_input % 3 == 0:
    print(user_input, "은(는)", "3의 배수다.")
else:
    remainder = user_input % 3
    print(user_input, "은(는)", "3으로 나누면 나머지가", remainder, "이다.")

입력한 값: 24
24 은(는) 3의 배수다.


25를 입력하면 다르게 출력된다.

In [ ]:
user_input = int(input("정수 하나를 입력하세요: "))
print("입력한 값:", user_input)

if user_input % 3 == 0:
    print(user_input, "은(는)", "3의 배수다.")
else:
    remainder = user_input % 3
    print(user_input, "은(는)", "3으로 나누면 나머지가", remainder, "이다.")

입력한 값: 25
25 은(는) 3으로 나누면 나머지가 1 이다.


**문제 2**

입력받은 두 부동소수점의 나눗셈을 아래 지정된 형식으로 출력하는 코드를 작성하라.
입력된 값은 각각 변수 `a`와 `b`에 저장한다.
예를 들어 3.3과 5.8이 입력되면 아래처럼 화면에 출력한다.
또한 계산 결과는 소수점 이하 셋째 자리에서 반올림한다.

    3.3 / 5.8 = 0.57

답:

부동소수점 두 개를 각각 입력받을 수 있도록 `input()` 함수를 `float()` 함수와 함께 사용한다.
아래 코드를 실행 한 후에 3.3과 5.8을 따로따로 입력하면 언급한 방식으로 화면에 결과가 출력된다.

In [ ]:
a = float(input("첫째 정수를 입력하세요: "))
print("첫째 정수:", a)
b = float(input("둘째 정수를 입력하세요: "))
print("둘째 정수:", b)

print(a, "/", b, "=", round(a / b, 2))

첫째 정수: 3.3
둘째 정수: 5.8
3.3 / 5.8 = 0.57


**문제 3**

아래 조건을 만족시키는 `taking_seats()` 함수를 정의하라.

- 매개변수는 사용하지 않는다.
- `taking_seats()` 형식으로 함수 호출이 일어나면 먼저 변수 `N` 지역변수에 할당할 양의 정수를 사용자로부터 입력받는다.
- `N` 변수가 가리키는 정수는 대기실에 일렬로 일렬로 놓여 있는 의자의 개수라 가정한다.
- 반환값은 없다. 대신 함수가 종료하기 전에 아래 조건을 만족하는 두 개의 문자열을 화면에 출력해야 한다.
    - 어느 두 사람도 이웃하여 앉을 수 없다는 조건하에, 앉을 수 있는 최대 인원이 앉는 방식과 최소 인원이 앉는 방식을 표현하는 문자열
    - 단, 자리를 채운 후에는 더 이상 어느 누구도 자리에 앉을 수 없는 방식이어야 함.

예를 들어 의자가 8개인 경우 아래 방식으로 앉는 것은 허용되지 않는다.
아래 표에서 `O`은 채워진 자리를, `X`는 빈자리를 가리킨다.

| 불가능 방식 | 이유 |
| :---: | :--- |
| `XXOXOXOX` | 왼쪽 끝에 한 명 더 앉을 수 있음 |
| `XOXXOXXX` | 오른쪽 끝에 한 명 더 앉을 수 있음|

의자가 8 개일 때 의자에 앉을 수 있는 최소 인원과 최대 인원은 다음과 같다.

| 인원 | 앉는 방식 |
| :---: | :---: |
| 최소 3명 | `XOXXOXXO` |
| 최대 4명 | `OXOXOXOX` |

따라서 함수가 호출되고 8을 입력하면 화면에 아래와 같이 출력되어야 한다.

```
입력한 값: 8
최대 인원 4 명: OXOXOXOX
최소 인원 3 명: XOXXOXXO
```

답:

`N`개의 의자가 있을 때 예를 들어 맨 왼쪽 의자에 앉은 다음에 한 자리 건널 때마다 앉으면 최대 인원이 앉게 된다.
즉 두 의자당 하나에 앉으면 된다. 만약 `N`이 홀수이면 맨 오른쪽 자리에 한 명이 더 앉을 수 있다.
따라서 `N`이 2배수인지 여부가 중요하다.

- `N`이 짝수일 때: 최대 인원은 `N//2`
- `N`이 홀수일 때: 최대 인원은 `N//2 + 1`

반면에 최소 인원은 다음과 같이 계산할 수 있다. 

먼저 의자가 1개 또는 2개이면 최대 1명 앉을 수 있다.
이제부터 의자가 3개 이상이라고 가정한다.
먼저 `N = 3`이라고 하자. 그러면 아래 방식으로 1명만 앉을 수 있다.

    XOX

`N > 3`인 경우는 3배수 여부로 결정된다.

- `N`이 3의 배수인 경우: `XOX` 방식을 `N//3` 번 반복한다.
    예를 들어 `N = 6`이면 아래 방식으로 최소 인원 `6//3` 명이 앉는다.

        XOXXOX
        
- `N`을 3으로 나눴을 때 나머지가 1인 경우: `XOX` 방식을 `N//3` 번 반복한 다음 나머지 한 자리에 앉는다.
    예를 들어 `N = 7`이면 아래 방식으로 최소 인원 `7//3` 명이 앉는다.

        XOXXOXO
                
- `N`을 3으로 나눴을 때 나머지가 2인 경우: `XOX` 방식을 `N//3` 번 반복한 다음 나머지 하나 또는 두 자리 중에 한 자리에 앉는다.
    예를 들어 `N = 8`이면 아래 방식으로 최소 인원 `8//3 + 1` 명이 앉는다.

        XOXXOXXO
        XOXXOXOX
        
    반면에 아래 방식은 한 명이 더 앉을 수 있기에 허용되지 않는다.
    
        XOXXOXXX

위 설명을 정리하여 코드로 작성하면 다음과 같다.

In [ ]:
def taking_seats():
    N = int(input("정수를 입력하세요: "))
    print("입력한 값:", N)

    # 최대 인원
    if (N%2 == 0) :
        max_num = N//2
        print("최대 인원", max_num, "명:", "OX" * max_num)
    else :
        max_num = (N//2) + 1
        print("최대 인원", max_num, "명:", "OX" * (max_num-1)+"O")

    # 최소 인원
    if (N%3 == 0) :
        min_num = N//3
        print("최소 인원", min_num, "명:", "XOX" * (min_num))
    else:
        min_num = (N//3) + 1
        if (N%3 == 1):
            print("최소 인원", min_num, "명:", "XOX" * (min_num-1)+ "O")
        else:
            print("최소 인원", min_num, "명:", "XOX" * (min_num-1) + "XO")

- 의자 8개

In [ ]:
taking_seats()

입력한 값: 8
최대 인원 4 명: OXOXOXOX
최소 인원 3 명: XOXXOXXO


- 의자 10개

In [ ]:
taking_seats()

입력한 값: 10
최대 인원 5 명: OXOXOXOXOX
최소 인원 4 명: XOXXOXXOXO


- 의자 20개

In [ ]:
taking_seats()

입력한 값: 20
최대 인원 10 명: OXOXOXOXOXOXOXOXOXOX
최소 인원 7 명: XOXXOXXOXXOXXOXXOXXO


## 람다 함수

### 연습문제

**문제 1**

평수를 제곱미터로 변환하는 코드를 람다 함수를 이용하여 작성하라. 
단, 25평을 제곱미터로 변환해야 하며,
소수점 이하 둘째자리까지만 계산한다.
평을 제곱미터로 변환하는 공식은 다음과 같다.

```python
평수 * 3.3058
```

답:

평수를 제곱미터로 변환하는 람다 함수는 다음과 정의할 수 있다.

In [ ]:
lambda pyeong: pyeong * 3.3058

<function __main__.<lambda>(pyeong)>

아래 코드는 이 람다 함수를 이용하여 평을 제곱미터로 변환한 결과를 출력한다.

In [ ]:
p = 25
square_meters = (lambda pyeong: pyeong * 3.3058)(p)

print("평수:", p, "평")
print("제곱미터:", round(square_meters, 2), "제곱미터")

평수: 25 평
제곱미터: 82.64 제곱미터


**문제 2**

제곱미터를 평으로 변환하는 람다 함수를 `lambda_pyeong` 변수에 할당한 다음에
그 변수를 이용하여 90 제곱미터가 몇 평이 되는지 계산하는 코드를 작성하라.
단, 소수점 이하 둘째 자리까지만 계산한다.

답:

제곱미터를 평으로 변환하는 람다 함수는 다음과 같다.

In [ ]:
lambda meters: meters / 3.3058

<function __main__.<lambda>(meters)>

아래 코드는 위 람다 함수를 `lambda_pyeong` 변수에 할당한 다음 80 제곱미터를 평으로 계산한 결과를 출력한다.

In [ ]:
sq = 80
lambda_pyeong = lambda meters: meters / 3.3058
pyeong = lambda_pyeong(sq)

print("제곱미터:", sq, "제곱미터")
print("평수:", round(pyeong, 2), "평")

제곱미터: 80 제곱미터
평수: 24.2 평


**문제 3**

함수를 인자로 사용하는 함수에 대한 문제다.

(1) 아래 `apply_chain()` 함수의 시그니처와 기능을 설명하라.

In [ ]:
def apply_chain(func1, func2, value):
    return func2(func1(value))

답:

함수 정의에 사용된 매개변수와 반환값은 다음과 같다.

- `func1`: 하나의 인자를 받는 함수를 인자로 받음
- `func2`: 하나의 인자를 받는 함수를 인자로 받음
- `value`: `func1`의 인자로 지정된 함수의 인자로 사용될 수 있는 값을 인자로 받음
- 반환값: `value`에 `func1`을 먼저 적용한 뒤, 그 결과에 `func2`를 적용한 값

(2) 아래 코드의 실행 결과를 설명하라.

In [ ]:
print(apply_chain(lambda c: c * 9 / 5 + 32, lambda f: round(f, 1), 26.3))

79.3


답:

`apply_chain()` 함수의 세 인자는 다음과 같다.
참고로 `func1`는 섭씨를 화씨로 변환하는 함수를,
`func2`는 소수점 둘째자리에서 반올림하는 함수를 가리킨다.

- `func1 = lambda c: c * 9 / 5 + 32`
- `func2 = lambda f: round(f, 1)`
- `value=26`

함수 본문의 실행 순서는 다음과 같다.

- `func1(26)` 호출: `26.3 * 9 / 5 + 32 = 79.34`  
- `func2(79.34)` 호출: `round(79.34, 1) = 79.3`
- 출력: `79.3`